# First Steps in Data Preprocessing

### What is data preprocessing?

Data preprocessing is the step in between cleaning up/exploring your data and building a machine learning model. The lines are often blurred; you may need to do some cleaning during preprocessing, and you might need to go back to preprocessing after modeling to tweak your feature set. Once you understand your dataset, you'll probably have some idea about how you want to model your data. Machine learning models in Python require numerical input, so if your dataset has categorical variables, you'll need to transform them. 

### Pandas refresher

To start, let's do a quick Pandas refresher (you should already be familiar with Pandas for this workshop). Pandas is a data processing library in Python that is incredibly useful for preprocessing data and is one of my all-time favorite Python libraries. 

Reading a dataset into pandas is pretty straightforward, and it's possible to read in data in a variety of formats. To read in a JSON file, we can use `read_json`. To check our results, let's look at the `head()` of the DataFrame:

In [1]:
import pandas as pd

# The path to the data directory. Easier than typing it out every time!
dir_string = "../../datasets/"

hiking = pd.read_json(dir_string + "hiking.json")
hiking.head()

Prop_ID                     Name  \
0    B057  Salt Marsh Nature Trail   
1    B073                Lullwater   
2    B073                  Midwood   
3    B073                Peninsula   
4    B073                Waterfall   

                                            Location      Park_Name  \
0  Enter behind the Salt Marsh Nature Center, loc...    Marine Park   
1  Enter Park at Lincoln Road and Ocean Avenue en...  Prospect Park   
2  Enter Park at Lincoln Road and Ocean Avenue en...  Prospect Park   
3  Enter Park at Lincoln Road and Ocean Avenue en...  Prospect Park   
4  Enter Park at Lincoln Road and Ocean Avenue en...  Prospect Park   

       Length Difficulty                                      Other_Details  \
0   0.8 miles       None  <p>The first half of this mile-long trail foll...   
1    1.0 mile       Easy  Explore the Lullwater to see how nature thrive...   
2  0.75 miles       Easy  Step back in time with a walk through Brooklyn...   
3   0.5 miles       Easy  Discover how the Peninsula has changed over th...   
4   0.5 miles       Easy  Trace the source of the Lake on the Waterfall ...   

  Accessible Limited_Access  lat  lon  
0          Y              N  NaN  NaN  
1          N              N  NaN  NaN  
2          N              N  NaN  NaN  
3          N              N  NaN  NaN  
4          N              N  NaN  NaN

Next, we'll read in a CSV of data on different wine attributes using `read_csv`, and once again check our input:

In [2]:
wine = pd.read_csv(dir_string + "wine_types.csv")
wine.head()

Type  Alcohol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0     1    14.23        1.71  2.43               15.6        127   
1     1    13.20        1.78  2.14               11.2        100   
2     1    13.16        2.36  2.67               18.6        101   
3     1    14.37        1.95  2.50               16.8        113   
4     1    13.24        2.59  2.87               21.0        118   

   Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

It's useful to be able to generate a list of the features present in your dataset. To see the columns in a datasets, use the `.columns` attribute:

In [3]:
hiking.columns

Index(['Prop_ID', 'Name', 'Location', 'Park_Name', 'Length', 'Difficulty',
       'Other_Details', 'Accessible', 'Limited_Access', 'lat', 'lon'],
      dtype='object')

And to see the data types of the columns in your dataset, use `.dtypes`:

In [4]:
hiking.dtypes

Prop_ID            object
Name               object
Location           object
Park_Name          object
Length             object
Difficulty         object
Other_Details      object
Accessible         object
Limited_Access     object
lat               float64
lon               float64
dtype: object

One of the most useful basic features of pandas is the ability to quickly output characteristics of your dataset like the mean, standard deviation, and quartile values. To see these values, you can `.describe()` your dataset. Note that this method will only work on continuous data.

In [5]:
wine.describe()

Type     Alcohol  Malic acid         Ash  Alcalinity of ash  \
count  178.000000  178.000000  178.000000  178.000000         178.000000   
mean     1.938202   13.000618    2.336348    2.366517          19.494944   
std      0.775035    0.811827    1.117146    0.274344           3.339564   
min      1.000000   11.030000    0.740000    1.360000          10.600000   
25%      1.000000   12.362500    1.602500    2.210000          17.200000   
50%      2.000000   13.050000    1.865000    2.360000          19.500000   
75%      3.000000   13.677500    3.082500    2.557500          21.500000   
max      3.000000   14.830000    5.800000    3.230000          30.000000   

        Magnesium  Total phenols  Flavanoids  Nonflavanoid phenols  \
count  178.000000     178.000000  178.000000            178.000000   
mean    99.741573       2.295112    2.029270              0.361854   
std     14.282484       0.625851    0.998859              0.124453   
min     70.000000       0.980000    0.340000              0.130000   
25%     88.000000       1.742500    1.205000              0.270000   
50%     98.000000       2.355000    2.135000              0.340000   
75%    107.000000       2.800000    2.875000              0.437500   
max    162.000000       3.880000    5.080000              0.660000   

       Proanthocyanins  Color intensity         Hue  \
count       178.000000       178.000000  178.000000   
mean          1.590899         5.058090    0.957449   
std           0.572359         2.318286    0.228572   
min           0.410000         1.280000    0.480000   
25%           1.250000         3.220000    0.782500   
50%           1.555000         4.690000    0.965000   
75%           1.950000         6.200000    1.120000   
max           3.580000        13.000000    1.710000   

       OD280/OD315 of diluted wines      Proline  
count                    178.000000   178.000000  
mean                       2.611685   746.893258  
std                        0.709990   314.907474  
min                        1.270000   278.000000  
25%                        1.937500   500.500000  
50%                        2.780000   673.500000  
75%                        3.170000   985.000000  
max                        4.000000  1680.000000

### Missing data

Now that we've refreshed ourselves on the basics of pandas, let's cover some ways for how to deal with missing data. There are a variety of techniques you can use, but here we'll focus on a few ways to remove missing data on a simple dataset, and then try out a few techniques on a couple of "real world" datasets in the next step.

Let's create a simple toy dataset with some missing data. Here we'll encode some `NaN` values into the dataset using NumPy and easily create a DataFrame with pandas's `DataFrame` method.

In [6]:
import numpy as np

df_dict = {"A": [6, 4, 9, np.NaN, 3],
           "B": [np.NaN, 8 , np.NaN, 2, 1],
           "C": [9, 3, np.NaN, np.NaN, 9]}

df = pd.DataFrame(df_dict)
df

A    B    C
0  6.0  NaN  9.0
1  4.0  8.0  3.0
2  9.0  NaN  NaN
3  NaN  2.0  NaN
4  3.0  1.0  9.0

You can drop all rows with missing data in your dataset with `.dropna()`:

In [7]:
df.dropna()

A    B    C
1  4.0  8.0  3.0
4  3.0  1.0  9.0

You can drop specific rows by passing index labels to the drop function, which defaults to dropping rows, using the `.drop()` method. 

In [8]:
df.drop([1, 2, 3])

A    B    C
0  6.0  NaN  9.0
4  3.0  1.0  9.0

Usually you'll want to focus on dropping a particular column, especially if all or most of its values are missing. If you want to drop columns, you can pass either a single column or a list of columns into the `.drop()` method. You'll also have to specify `axis=1` in order to drop columns instead of rows.

In [9]:
df.drop("A", axis=1)

B    C
0  NaN  9.0
1  8.0  3.0
2  NaN  NaN
3  2.0  NaN
4  1.0  9.0

What if we want to drop rows where data is missing in a particular column? We can do this with the help of boolean indexing, which is a way to filter a DataFrame through a conditional statement. Here we'll select all rows from the dataset where column C is equal to 9:

In [10]:
df[df["C"] == 9]

A    B    C
0  6.0  NaN  9.0
4  3.0  1.0  9.0

Finally, we can use boolean indexing in combination with the `.notnull()` method to select all rows in a dataset where a certain column or set of columns is not null.

In [11]:
df[df["C"].notnull()]

A    B    C
0  6.0  NaN  9.0
1  4.0  8.0  3.0
4  3.0  1.0  9.0

### Your turn!

We have a dataset comprised of volunteer information from New York City. We can think of each column in the dataset as a feature. This dataset has features pertaining to the time of the volunteering event, the category, the location, the organization, and so on. 

The dataset has a number of features, but let's say that we want to get rid of features that have no more than 65 missing values. 

First, let's read in the dataset and take a look at its `head()`:

In [12]:
volunteer = pd.read_csv(dir_string + "volunteer.csv")
volunteer.head()

opportunity_id  content_id  vol_requests  event_time  \
0            4996       37004            50           0   
1            5008       37036             2           0   
2            5016       37143            20           0   
3            5022       37237           500           0   
4            5055       37425            15           0   

                                               title  hits  \
0  Volunteers Needed For Rise Up & Stay Put! Home...   737   
1                                       Web designer    22   
2      Urban Adventures - Ice Skating at Lasker Rink    62   
3  Fight global hunger and support women farmers ...    14   
4                                      Stop 'N' Swap    31   

                                             summary is_priority  category_id  \
0  Building on successful events last summer and ...         NaN          NaN   
1             Build a website for an Afghan business         NaN          1.0   
2  Please join us and the students from Mott Hall...         NaN          1.0   
3  The Oxfam Action Corps is a group of dedicated...         NaN          1.0   
4  Stop 'N' Swap reduces NYC's waste by finding n...         NaN          4.0   

               category_desc  ...     end_date_date    status Latitude  \
0                        NaN  ...      July 30 2011  approved      NaN   
1  Strengthening Communities  ...  February 01 2011  approved      NaN   
2  Strengthening Communities  ...   January 29 2011  approved      NaN   
3  Strengthening Communities  ...     March 31 2012  approved      NaN   
4                Environment  ...  February 05 2011  approved      NaN   

   Longitude  Community Board Community Council  Census Tract  BIN  BBL NTA  
0        NaN              NaN                NaN          NaN  NaN  NaN NaN  
1        NaN              NaN                NaN          NaN  NaN  NaN NaN  
2        NaN              NaN                NaN          NaN  NaN  NaN NaN  
3        NaN              NaN                NaN          NaN  NaN  NaN NaN  
4        NaN              NaN                NaN          NaN  NaN  NaN NaN  

[5 rows x 35 columns]

Let's take a look at the shape of the dataset as well:

In [13]:
volunteer.shape

(665, 35)

So, we have 665 rows and 35 columns. We're going to drop those columns that have more than 65 missing values. We can do this with the `.dropna()` method by setting two parameters: `axis=1` to specify we want to drop columns, which you've already done above, and `thresh=600` to specify that we want to keep columns with at least 600 non-null values. 

Fill in the parameters below:

In [14]:
volunteer_drop_columns = volunteer.dropna(axis=1, thresh=600)
volunteer_drop_columns.shape

(665, 22)

As you can see, we've dropped 13 columns that had at least 65 missing values (and 600 non-null values).

Taking a look at this dataset again, we want to drop rows where the `category_desc` column values are missing. We're going to do this using boolean indexing, by checking to see if we have any null values, and then filtering the dataset so that we only have rows with those values.

First, we can check how many values are missing in a column using `.isnull()` to get the null values and `.sum()` to get the total:

In [15]:
volunteer["category_desc"].isnull().sum()

48

So, we have 48 null values. We can once again use `.notnull()` to subset the DataFrame using boolean indexing:

In [16]:
volunteer_subset = volunteer[volunteer["category_desc"].notnull()]

volunteer_subset.shape

(617, 35)

You can see that we've retained all 35 features and that we've dropped the 48 rows where the column `category_desc` didn't have a value.

### Exploring data types

One of the next steps of preprocessing is to think about the types that are present in your dataset, because you'll likely have to transform some of these columns to other types later on. Let's take a deeper look at types as well as how to convert column types in your dataset.

You can check the types of a DataFrame by using the `.dtypes` attribute:

In [17]:
volunteer.dtypes

opportunity_id          int64
content_id              int64
vol_requests            int64
event_time              int64
title                  object
hits                    int64
summary                object
is_priority            object
category_id           float64
category_desc          object
amsl                  float64
amsl_unit             float64
org_title              object
org_content_id          int64
addresses_count         int64
locality               object
region                 object
postalcode            float64
primary_loc           float64
display_url            object
recurrence_type        object
hours                   int64
created_date           object
last_modified_date     object
start_date_date        object
end_date_date          object
status                 object
Latitude              float64
Longitude             float64
Community Board       float64
Community Council     float64
Census Tract          float64
BIN                   float64
BBL       

Here's a list of some of the most common datatypes:
- `object`: pandas uses this type to refer to a column that consists of string values or is of mixed types. 
- `int64`: the equivalent of the Python integer type. The 64 simply refers to memory allotted for storing the values. 
- `float64`: the equivalent of the Python float type. 
- `datetime64` (or the `timedelta` type): dates in pandas can be stored as a special type that makes time operations easier or set as an index.

Sometimes, you'll start working with a dataset that has an incorrect column type: maybe a numerical column was written out into a csv as a string, and when you try to work with that column, numerical operations won't work. 

Let's convert some columns, first using a small example dataset:


In [18]:
df_types = pd.DataFrame({"A": [1, 2, 3], 
                         "B": ["string1", "string2", "string3"], 
                         "C": ["1.0", "2.0", "3.0"]})

print(df_types)
print("\n")
print(df_types.dtypes)

   A        B    C
0  1  string1  1.0
1  2  string2  2.0
2  3  string3  3.0


A     int64
B    object
C    object
dtype: object


As you can see, column C has the `dtype` of `object`. However, looking at it, it's clear that column C should be the type `float`. Converting a column type is pretty straightforward using `.astype()`:

In [19]:
df_types["C"] = df_types["C"].astype("float")

df_types.dtypes

A      int64
B     object
C    float64
dtype: object

Now column C is the correct type.

### Class imbalance and distribution

One of the most necessary steps for preprocessing, which you should be familiar with if you've taken other courses on Python and machine learning, is splitting up your data into training and test sets. We do this to avoid the issue of overfitting. If we train a model on our entire set of data, we won't have any way to test and validate our model because the model will essentially know the dataset by heart. Holding out a test set allows us to preserve some data the model hasn't seen yet.

Splitting data into training and test sets is straightfoward with scikit-learn's `train_test_split()` function. This example, on a simple dataset, should be familiar already:

In [20]:
from sklearn.model_selection import train_test_split

split_df = pd.DataFrame({"X": [1, 2, 3, 4, 5, 6], "y": ["y", "y", "n", "n", "y", "n"]})

print(split_df)

X_train, X_test, y_train, y_test = train_test_split(split_df["X"], split_df["y"])

   X  y
0  1  y
1  2  y
2  3  n
3  4  n
4  5  y
5  6  n


Let's take a look at the split sets using `pd.concat()`

In [21]:
print("Training set")
print(pd.concat([X_train, y_train], axis=1))

print("\n")

print("Test set")
print(pd.concat([X_test, y_test], axis=1))

Training set
   X  y
2  3  n
0  1  y
3  4  n
5  6  n


Test set
   X  y
1  2  y
4  5  y


This method works well, but there are going to be scenarios where your `y` values—the values you're potentially predicting in a model—aren't as evenly distributed in your dataset. For example, if 80% of your dataset falls into the `y` class, and 20% falls into the `n` class, it's possible that when you run `train_test_split`, this distribution isn't represented in your test set. Possibly you'll end up with no `n` values in the test set, which will severely bias your model.

To avoid this scenario, we can use a technique called stratified sampling, which is a way of sampling that takes into account the distribution of classes or features in your dataset. With stratified sampling, we can preserve that 80%/20% distribution across both our training and test sets, which means we're training and testing our model in similar circumstances.

First, let's read in the example dataset and look at the distribution for our `Y` column:

In [22]:
stratified_df = pd.read_csv(dir_string + "stratified_df.csv")

print(stratified_df.head())
print("\n")
print(stratified_df["Y"].value_counts())

   X  Y
0  1  n
1  2  y
2  3  y
3  4  n
4  5  y


y    80
n    20
Name: Y, dtype: int64


So here we have 80% of our data with a `y` label, and 20% with `n`. In order to preserve this distribution when splitting up the dataset, we can use the `stratify=` parameter in the `train_test_split` function. All we need to do is pass in the values we want to stratify the split by, which in this case is the `Y` column.

In [23]:
X_train, X_test, y_train, y_test = train_test_split(stratified_df["X"], stratified_df["Y"], stratify=stratified_df["Y"])

print(y_train.value_counts())
print("\n")
print(y_test.value_counts())

y    60
n    15
Name: Y, dtype: int64


y    20
n     5
Name: Y, dtype: int64


You can see the distribution of classes is in accordance with the original y class distribution. 

### Your turn!

Let's say that, using the `volunteer` dataset, we're thinking about trying to predict the `category_desc` variable using the other features in the dataset. First, though, we need to know what the class distribution (and imbalance) is for that particular label. We're going to use the `volunteer` dataset without nulls in the `category_desc` column, which we created earlier in the notebook.

First, let's check the distribution using `.value_counts()`:

In [24]:
volunteer_subset["category_desc"].value_counts()

Strengthening Communities    307
Helping Neighbors in Need    119
Education                     92
Health                        52
Environment                   32
Emergency Preparedness        15
Name: category_desc, dtype: int64

Across the different categories, we have an uneven distribution of classes. If we were building a model to predict these classes, we'd want to make sure we retain this distribution across the training and test sets. So, let's prepare the `volunteer` dataset for stratified sampling. Since the `volunteer` dataset has so many features, we're going to break out the `category_desc` to use as our predicted class:

In [25]:
volunteer_X = volunteer_subset.drop("category_desc", axis=1)
volunteer_y = volunteer_subset["category_desc"]

Now we can pass the data into `train_test_split()` and set the `stratify=` parameter to the `category_desc` distribution:

In [26]:
X_train, X_test, y_train, y_test = train_test_split(volunteer_X, volunteer_y, stratify=volunteer_y)

print(y_train.value_counts())
print("\n")
print(y_test.value_counts())

Strengthening Communities    230
Helping Neighbors in Need     89
Education                     69
Health                        39
Environment                   24
Emergency Preparedness        11
Name: category_desc, dtype: int64


Strengthening Communities    77
Helping Neighbors in Need    30
Education                    23
Health                       13
Environment                   8
Emergency Preparedness        4
Name: category_desc, dtype: int64


Both sets have the original distribution preserved.